dynatrace_topology/stackstate_checks/dynatrace_topology/dynatrace_topology.py

```diff
@@ -0,0 +1,276 @@
+        self.domain = instance.get('domain', '')
+        self.environment = instance.get('environment', 'production')
+        self.verify = instance.get('verify', True)
+        self.cert = instance.get('cert', '')
+        self.keyfile = instance.get('keyfile', '')
+
+        try:
+            self.start_snapshot()
+            self.process_topology()
+            self.stop_snapshot()
+            msg = "Dynatrace topology processed successfully"
+            self.service_check(self.SERVICE_CHECK_NAME, AgentCheck.OK, tags=self.tags, message=msg)
+        except Exception as e:
+            self.log.exception(str(e))
+            self.service_check(self.SERVICE_CHECK_NAME, AgentCheck.CRITICAL, tags=self.tags, message=str(e))
+
+    def process_topology(self):
+        """
+        Collects components and relations for each component type from dynatrace smartscape topology API
+        """
+        start_time = datetime.now()
+        self.log.info("Starting the collection of topology")
+        services = self.collect_services()
+        processes = self.collect_processes()
+        processes_groups = self.collect_process_groups()
+        hosts = self.collect_hosts()
+        applications = self.collect_applications()
+        topology = {"service": services, "process": processes, "process-group": processes_groups, "host": hosts,
+                    "application": applications}
+        # process component for each type
+        for comp_type, response in topology.items():
+            self.process_component(response, comp_type)
+        end_time = datetime.now()
+        time_taken = end_time - start_time
+        self.log.info("Time taken to collect the topology is: {} seconds".format(time_taken.total_seconds()))
+
+    def collect_relations(self, component, externalId):
+        """
+        Collects relationships from different component-types
+        :param component: the component for which relationships need to be extracted and processed
+        :param externalId: the component externalId for and from relationship will be created
+        :return: None
+        """
+        outgoing_relations = component.get("fromRelationships", {})
+        incoming_relations = component.get("toRelationships", {})
+        for relation_type in outgoing_relations.keys():
+            # Ignore `isSiteOf` relation since location components are not processed right now
+            if relation_type != "isSiteOf":
+                for target_id in outgoing_relations[relation_type]:
+                    self.relation(externalId, target_id, relation_type, {})
+        for relation_type in incoming_relations.keys():
+            # Ignore `isSiteOf` relation since location components are not processed right now
+            if relation_type != "isSiteOf":
+                for source_id in incoming_relations[relation_type]:
+                    self.relation(source_id, externalId, relation_type, {})
+
+    def filter_data(self, data):
```

Comment:
Can we make this a bit more generic and call it `unset`. Then we can just use it wherever we need with any key and it makes it really clear what we are removing. `del` mutates so it probably doesn't make sense to return data and assign it every time. So something like:

```
def unset(self, data, key):
  if key in data: del data[key]
```

Reply:
i made it for now `filter_item_topology_data` to remove all three in the same method.